### PR TITLE
Remove Oracle Database JDBC driver from the Keycloak distribution

### LINKED ISSUE #22452: Ask admins to install the Oracle Database driver separately
### Description

Due to the Oracle driver being proprietary and not open source, it shouldn't be shipped as part of the Keycloak distribution. 

As [Oracle recommends](https://www.oracle.com/database/technologies/appdev/jdbc-downloads.html), "The best practice is use the latest version of the Oracle JDBC driver that supports the JDK and Database version that you use and is compatible with your support requirements."

Admins will need to download and install the drivers separately, depending among others on the Oracle database version they are using.

### Discussion

_No response_

### Motivation

Keycloak should ship only open source database drivers.

### Details

I'll prepare a PR for this

### PR DESCRIPTION
**Pull request by mabartos**
Extended version for completion of requirements for 22.x release

Closes #22452

Requirements for 22.x backport:

- [x] The driver would remain in upstream included by default as it's a breaking change to do in a micro.
- [x] Introduce -DskipProprietary property to remove the driver. This would be used only by RHBK 22 pipelines. (Only the `product` profile is introduced; IMHO better approach)
- [x] Docs and testsuite would need to be adapted to reflect the different behaviour (driver included vs removed) between Keycloak and RHBK.

--------------------------
~~The `-DskipProprietary` property is present in the codebase, and moreover, it's possible to ignore the proprietary Oracle JDBC driver by specifying profile `-Pproduct`. It'll provide us better manageability around product builds as it will not be necessary to include all provided flags which distinguish the build approaches.~~

EDIT: Only the `product` profile is introduced.

The `-Pproduct` profile will need to be provided for other parts of Keycloak anyway. 

I'm not sure about the docs changes whether they should be first added to the `main` in a separate PR and then backported to 22 and used here. To be more specific, there will be a necessity for providing such a functionality even in later Keycloak releases. 

@vmuzikar @mhajas WDYT?

### CODE REVIEW

File: testsuite/integration-arquillian/servers/auth-server/quarkus/pom.xml
Comment (mabartos) [inline testsuite/integration-arquillian/servers/auth-server/quarkus/pom.xml:213]: We will add the Oracle JDBC driver to the `providers` folder of the dist even when the Oracle JDBC driver is present via transitive dependency of the Quarkus Oracle JDBC extension. We will avoid further differentiation between `main` and `22.x`. As the driver will be only part of 22.x, I don't see any benefit to making other workarounds bcs of testsuite.

@vmuzikar Is it ok for you?

File: docs/guides/templates/profile.adoc
Comment (mabartos) [inline docs/guides/templates/profile.adoc:11]: Such changes will be probably necessary to add even for the `docs/documentation`, but I don't want to provide it there if there's no use for it ATM.

File: docs/guides/server/db.adoc
Comment (mabartos) [inline docs/guides/server/db.adoc:31]: Installing the DB driver only when it's related to RHBK.

File: testsuite/integration-arquillian/servers/auth-server/quarkus/pom.xml
Comment (mhajas) [reply testsuite/integration-arquillian/servers/auth-server/quarkus/pom.xml:213]: If we add this to the `product` profile, would it have some disadvantages?

File: quarkus/runtime/pom.xml
Comment (mhajas) [reply quarkus/runtime/pom.xml:752]: I checked this and it works:
```
$ mvn dependency:tree | grep oracle
[INFO] +- io.quarkus:quarkus-jdbc-oracle:jar:3.2.4.Final:compile
[INFO] |  +- com.oracle.database.jdbc:ojdbc11:jar:23.2.0.0:compile
[INFO] |  \- com.oracle.database.nls:orai18n:jar:23.2.0.0:compile
```
vs
```
$ mvn dependency:tree -Pproduct | grep oracle
[INFO] +- io.quarkus:quarkus-jdbc-oracle:jar:3.2.4.Final:compile
```

File: quarkus/runtime/pom.xml
Comment (mhajas) [inline quarkus/runtime/pom.xml:735]: Is there some reason we need this property name? Can we just use `-Dproduct` or `-Pproduct` similarly as we used it before?

File: testsuite/integration-arquillian/servers/auth-server/quarkus/pom.xml
Comment (mabartos) [reply testsuite/integration-arquillian/servers/auth-server/quarkus/pom.xml:213]: @mhajas ahh, probably not. Better approach.

File: quarkus/runtime/pom.xml
Comment (mabartos) [reply quarkus/runtime/pom.xml:735]: I thought it could be helpful for the community builds as users could specify the property as it's more specific than the product profile. However, when I'm thinking about it right now, it could be sufficient to use only the `product` profile as the property was supposed to be an internal one anyway.

Comment (mabartos) [reply quarkus/runtime/pom.xml]: I've removed it.

File: testsuite/integration-arquillian/servers/auth-server/quarkus/pom.xml
Comment (mabartos) [inline testsuite/integration-arquillian/servers/auth-server/quarkus/pom.xml:341]: Works as expected. The Oracle JDBC driver is included in the `/providers` dir only when the `product` profile is activated.

**Review by mhajas — APPROVED** [review]
Thank you @mabartos! From my point of view LGTM. However, I would wait for a review from @vmuzikar as he was involved in more discussions around this.

File: docs/guides/.asciidoctorconfig
Comment (stianst) [inline docs/guides/.asciidoctorconfig:7]:
```suggestion
```

File: docs/guides/pom.xml
Comment (stianst) [inline docs/guides/pom.xml:106]:
```suggestion
```

File: docs/guides/pom.xml
Comment (stianst) [inline docs/guides/pom.xml:36]:
```suggestion
```

File: docs/guides/pom.xml
Comment (stianst) [inline docs/guides/pom.xml:184]:
```suggestion
```

File: docs/guides/server/db.adoc
Comment (stianst) [inline docs/guides/server/db.adoc:68]: I don't really like the inclusion of DockerFile examples in the guides. I'd suggest just removing this for now, and perhaps in the future we could add some example DockerFiles to QuickStarts repository? Will be easier to maintain and also to try out if it's an actual DockerFile in a repository rather than a code snippet in documentation.

File: docs/guides/templates/profile.adoc
Comment (stianst) [inline docs/guides/templates/profile.adoc:11]:
```suggestion
<#macro ifProduct>
ifdef::project_product[]
<#nested>
endif::[]
</#macro>

<#macro ifCommunity>
ifndef::project_product[]
<#nested>
endif::[]
</#macro>
```

File: testsuite/integration-arquillian/servers/auth-server/quarkus/pom.xml
Comment (stianst) [reply testsuite/integration-arquillian/servers/auth-server/quarkus/pom.xml:341]: This doesn't do anything really as we test DB support with GitHub Actions which doesn't use the `product` profile. Also not sure we should always install the Oracle Driver, but rather only install it when we are testing Oracle.

**Review by stianst — COMMENTED** [review]
Instead of adding another way of marking sections as community/product it should use the same approach as in `documentation` already. That is also already supported by the `product` documentation build, so these suggested changes works with downstream doc builds without any changes needed there.

I'm not too keen on including DockerFile examples in the documentation as they are hard and error prone to test and use. Would be better to add examples to `keycloak-quickstarts` and link to those, but I think we can also live without an actual DockerFile example perhaps.

Final comment is on how we can automate testing. The included profile in the testsuite doesn't really achieve anything as we are testing DB support in community only using GitHub Actions, so our testing will be done with the Oracle driver included in the dist, rather than added to providers directory. I think we'd have to introduce another job in ci.yml that re-builds the distribution without Oracle driver, then adds it to providers, and runs testing for Oracle. That'd be for `release/22.0` branch, for `main` when we remove Oracle driver fully we wouldn't need this additional job, but rather just install the Oracle driver into providers, but think we should only do that in the `oracle` matrix for `legacy-store-integration-tests` rather than having it always included as what will happen here.

**Comment by mabartos** [comment]
@stianst Thanks for the review

> Instead of adding another way of marking sections as community/product it should use the same approach as in documentation already. That is also already supported by the product documentation build, so these suggested changes works with downstream doc builds without any changes needed there.

Yep, better.

**Comment by mabartos** [comment]
> The included profile in the testsuite doesn't really achieve anything as we are testing DB support in community only using GitHub Actions, so our testing will be done with the Oracle driver included in the dist, rather than added to providers directory. I think we'd have to introduce another job in ci.yml that re-builds the distribution without Oracle driver, then adds it to providers, and runs testing for Oracle.

I'm not very familiar with the chosen approach to testing the product. We will not be able to avoid changing certain parts of Keycloak for the product (f.e. GELF support), so also the testing will be slightly different. I've kinda counted on providing the `product` profile even for the test execution when the product is tested, so we could asses the behavior of the testsuite.

However, in this case, it'd be fine not to use the `product` profile in the testsuite for these artifacts, but rather the profile that is turned on only when OracleDB is tested.

Introducing the new CI job might make sense to simulate the product behavior; Oracle JDBC driver not included.

File: docs/guides/server/db.adoc
Comment (vmuzikar) [reply docs/guides/server/db.adoc:68]: +1
However, if we decide to change this, we'd need to do it in `main` too as it's [there already](https://github.com/keycloak/keycloak/blob/660d17f8ec113558e520be97007f6b4a72f0fca6/docs/guides/server/db.adoc?plain=1#L55-L64) and this is just a backport. I'd suggest continuing as is and address this change as a follow-up. There are many more docker examples throughout the  guides, it's not limited to this one. @stianst WDYT?

**Comment by vmuzikar** [comment]
> Introducing the new CI job might make sense to simulate the product behavior

I'd argue we have internal CI pipelines for this.

File: docs/guides/templates/profile.adoc
Comment (vmuzikar) [inline docs/guides/templates/profile.adoc:1]: Haven't tested this but assuming it works this PR LGTM.

**Review by vmuzikar — APPROVED** [review]

File: docs/guides/templates/profile.adoc
Comment (mabartos) [reply docs/guides/templates/profile.adoc:1]: I can confirm it is working for the upstream and product docs.

**Comment by mabartos** [comment]
The approach for docs works as expected. I agree to handle the Dockerfiles in follow-up tasks. Moreover, as we still have the QE pipelines that will be used for the product, I'd say this approach might be sufficient; @vmuzikar good argument. 

@stianst @mhajas Are you ok with the current approach? 

@vmuzikar Could you please squash the commits through GH, or should I do it manually?

File: docs/guides/templates/profile.adoc
Comment (mhajas) [inline docs/guides/templates/profile.adoc:2]: In other places we use `ifeval::[{project_product}==true]`, do you think it would make sense to make it consistent?

**Review by mhajas — APPROVED** [review]
Thank you for the PR @mabartos! I added just one nitpick comment about using `ifdef` instead of `ifeval` as we do in other places. I am approving though as I can confirm it works as expected even with `ifdef` so. I am leaving it up to you whether you want to leave this as is or change it.

File: docs/guides/templates/profile.adoc
Comment (mabartos) [reply docs/guides/templates/profile.adoc]: Done

**Comment by mabartos** [comment]
@mhajas Done

**Review by mhajas — APPROVED** [review]
thanks @mabartos!

File: docs/guides/server/db.adoc
Comment (stianst) [reply docs/guides/server/db.adoc:68]: Oki, that works

**Review by stianst — APPROVED** [review]

File: docs/guides/server/db.adoc
Comment (vmuzikar) [reply docs/guides/server/db.adoc:68]: Created https://github.com/keycloak/keycloak/issues/22715

**Pull request merged**
